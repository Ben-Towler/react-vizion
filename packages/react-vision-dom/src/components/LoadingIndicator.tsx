import React from 'react';
import classnames from 'classnames';

type Props = {
  isLoading?: boolean;
};

const LoadingIndicator: React.FunctionComponent<Props> = ({ isLoading }) => {
  return (
    <div>
      <span className={classnames({ loading: isLoading })} />
    </div>
  );
};

export default LoadingIndicator;
