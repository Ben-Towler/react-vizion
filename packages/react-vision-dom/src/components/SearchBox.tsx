import React, { Component } from 'react';
import LoadingIndicator from './LoadingIndicator';
import { createClassNames } from '../core/utils';
import classnames from 'classnames';
import { translatable } from 'react-vision-core';

type Props = {
  currentRefinement?: string;
  searchAsYouType?: boolean;
  className?: string;
  placeholder?: string;
  style?: React.CSSProperties;

  autoFocus?: boolean;
  loading?: boolean;
  showLoadingIndicator?: boolean;
  disabled?: boolean;

  onClear?: (event?: any) => void;
  onFocus?: (event?: React.FocusEvent) => void;
  onBlur?: (event?: React.FocusEvent) => void;
  onInput?: (event: React.FormEvent<HTMLInputElement>) => void;
  onChange?: (event?: React.ChangeEvent<HTMLInputElement>) => void;
  onKeyPress?: (event?: React.KeyboardEvent<HTMLInputElement>) => void;
  onSubmit?: (event: any) => void;

  clearButton?: (clearSearch: () => void) => React.ReactNode;
  submitButton?: React.ReactNode;
  loadingIndicator?: React.ReactNode;
  translate: any;
};

type State = {
  query: string;
  data: any[];
};

const cx = createClassNames('SearchBox');

const mockData = [
  { id: '1', name: 'Clinic 1' },
  { id: '2', name: 'Clinic 2' },
  { id: '3', name: 'Clinic 3' },
  { id: '4', name: 'Clinic 4' },
  { id: '5', name: 'Clinic 5' },
  { id: '6', name: 'Clinic 6' },
  { id: '7', name: 'Clinic 7' },
  { id: '8', name: 'Clinic 8' },
  { id: '9', name: 'Clinic 9' },
  { id: '10', name: 'Clinic 10' },
];

class SearchBox extends Component<Props, State> {
  formRef;
  state = {
    query: '',
    data: [],
  };

  constructor(props) {
    super(props);
    // We bind functions for test purposes instead of using arrow functions
    this.onSubmit = this.onSubmit.bind(this);
    this.onInput = this.onInput.bind(this);
    this.onClear = this.onClear.bind(this);
    this.formRef = React.createRef();
  }

  onSubmit(event) {
    event.preventDefault();
    const { query } = this.state;

    const fetchedData = mockData.filter(data =>
      data.name.toLowerCase().includes(query.toLowerCase())
    );

    this.setState({ data: fetchedData });

    if (this.props.onSubmit) this.props.onSubmit(event);
  }

  onInput(event: React.ChangeEvent<HTMLInputElement>) {
    const query = event.target.value;
    const { searchAsYouType } = this.props;

    this.setState({ query }, () => {
      if (searchAsYouType) {
        this.formRef.dispatchEvent(new Event('submit'));
      }
    });

    if (this.props.onInput) this.props.onInput(event);
  }

  onClear() {
    const query = '';
    this.setState({ query });
  }

  render() {
    const {
      autoFocus,
      className,
      clearButton,
      disabled,
      loading,
      loadingIndicator,
      showLoadingIndicator,
      onBlur,
      onChange,
      onClear,
      onFocus,
      onKeyPress,
      placeholder,
      style,
      submitButton,
      translate,
    } = this.props;

    const { query } = this.state;

    return (
      <div className={classnames(cx(''), className)} style={style}>
        <form onSubmit={this.onSubmit} ref={ref => (this.formRef = ref)}>
          <div style={{ display: 'flex' }}>
            <input
              id="search-box-input"
              type="search"
              autoComplete="off"
              autoCorrect="off"
              autoCapitalize="off"
              disabled={disabled}
              spellCheck={false}
              className={cx('input')}
              onInput={this.onInput}
              onChange={onChange}
              onKeyPress={onKeyPress}
              onBlur={onBlur}
              autoFocus={autoFocus}
              onFocus={onFocus}
              value={query}
              placeholder={
                placeholder ? placeholder : translatable('placeholder')
              }
            />
            {showLoadingIndicator &&
              loading &&
              (loadingIndicator ? (
                loadingIndicator
              ) : (
                <LoadingIndicator isLoading={loading} />
              ))}
            {clearButton ? (
              clearButton(this.onClear)
            ) : (
              <button
                id="search-box-clear"
                className={cx('clear')}
                onClick={onClear ? onClear : this.onClear}
              >
                {translate('clearText')}
              </button>
            )}
            {submitButton ? (
              submitButton
            ) : (
              <button className={cx('submit')} type="submit">
                {translate('searchText')}
              </button>
            )}
          </div>
        </form>
      </div>
    );
  }
}

export default translatable({
  placeholder: 'Default',
  searchText: 'Search',
  clearText: 'Clear',
})(SearchBox);
