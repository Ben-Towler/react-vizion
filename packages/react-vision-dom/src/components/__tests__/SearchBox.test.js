import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import SearchBox from '../SearchBox';
import LoadingIndicator from '../LoadingIndicator';

Enzyme.configure({ adapter: new Adapter() });

describe('SearchBox', () => {
  it('applies its default props', () => {
    const instance = renderer.create(<SearchBox refine={() => null} />);

    expect(instance.toJSON()).toMatchSnapshot();

    instance.unmount();
  });

  it('should have focus after render when autoFocus prop is true', () => {
    //Creates a parent div where the component will be attached to
    //so that way we can get the active element id
    document.body.innerHTML = '<div></div>';

    const wrapper = mount(<SearchBox autoFocus={true} />, {
      attachTo: document.getElementsByName('div')[0],
    });

    const element = wrapper.find('input');
    expect(element.props().id).toEqual(document.activeElement.id);
  });

  it('should initialize state with empty query', () => {
    const wrapper = shallow(<SearchBox />);
    expect(wrapper.state('query')).toEqual('');
  });

  describe('Style tests', () => {
    it('should render main div with className props', () => {
      const classNameTest = 'class-test';
      const wrapper = shallow(<SearchBox className={classNameTest} />);

      expect(wrapper.first('div').hasClass(classNameTest)).toBeTruthy();
    });

    it('should render main div with style props ', () => {
      const styleTest = { top: 0 };
      const wrapper = shallow(<SearchBox style={styleTest} />);

      expect(wrapper.first('div').props().style).toEqual(styleTest);
    });
  });

  describe('Custom inner components', () => {
    it('should render with custom loading indicator', () => {
      const customLoadingIndicator = <span>Loading</span>;
      const wrapper = shallow(
        <SearchBox
          loadingIndicator={customLoadingIndicator}
          showLoadingIndicator={true}
          loading={true}
        />
      );
      const defaultLoadingIndicator = <LoadingIndicator />;

      expect(wrapper.contains(customLoadingIndicator)).toBeTruthy();
      expect(wrapper.contains(defaultLoadingIndicator)).toBeFalsy();
    });

    it('should render with custom clear button', () => {
      const customClearButton = <button>Test</button>;
      const wrapper = shallow(
        <SearchBox clearButton={_ => customClearButton} />
      );

      expect(wrapper.contains(customClearButton)).toBeTruthy();
      expect(wrapper.exists('#search-box-clear')).toBeFalsy();
    });

    it('should render with custom submit button', () => {
      const customSubmitButton = <button>Test</button>;
      const wrapper = shallow(<SearchBox submitButton={customSubmitButton} />);

      expect(wrapper.contains(customSubmitButton)).toBeTruthy();
      expect(wrapper.exists('.search-box-button')).toBeFalsy();
    });
  });

  describe('Event tests', () => {
    describe('Default events', () => {
      it('should update state on default input change', () => {
        const wrapper = shallow(<SearchBox />);
        const enteredInputValue = 'a';

        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(wrapper.state('query')).toEqual(enteredInputValue);
      });

      it('should not submit on default input change when "searchAsYouType" prop is not valid ', () => {
        const wrapper = shallow(<SearchBox />);
        const enteredInputValue = 'a';
        const spyOnSubmit = spyOn(wrapper.instance(), 'onSubmit');

        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(wrapper.state().query).toEqual(enteredInputValue);
        expect(spyOnSubmit).not.toHaveBeenCalled();
      });

      it('should submit on default input change when "searchAsYouType" prop is true ', () => {
        const wrapper = mount(<SearchBox searchAsYouType={true} />);
        const enteredInputValue = 'a';
        const spyOnSubmit = spyOn(wrapper.instance(), 'onSubmit');

        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(wrapper.state().query).toEqual(enteredInputValue);
        expect(spyOnSubmit).toHaveBeenCalled();
      });
      it('should submit on button submit', () => {
        const spyOnSubmit = spyOn(SearchBox.prototype, 'onSubmit');
        const wrapper = mount(<SearchBox />);

        wrapper.find('button').simulate('submit');

        expect(spyOnSubmit).toHaveBeenCalled();
      });
      it('should clear state on default clear search click', () => {
        const wrapper = shallow(<SearchBox />);
        const enteredInputValue = 'a';

        //Simulate input to change the default state query value
        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(wrapper.state().query).toEqual(enteredInputValue);
        wrapper.find('#search-box-clear').simulate('click');
        expect(wrapper.state().query).toEqual('');
      });
    });

    describe('Custom events', () => {
      it('should trigger onBlur on input blur', () => {
        const onBlur = jest.fn();

        const wrapper = shallow(<SearchBox onBlur={onBlur} />);
        wrapper.find('input').simulate('blur');

        expect(onBlur).toHaveBeenCalled();
      });
      it('should trigger onChange on input change', () => {
        const onChange = jest.fn();

        const wrapper = shallow(<SearchBox onChange={onChange} />);
        wrapper.find('input').simulate('change');

        expect(onChange).toHaveBeenCalled();
      });
      it('should trigger onClear on input change', () => {
        const onClear = jest.fn();

        const wrapper = shallow(<SearchBox onClear={onClear} />);
        wrapper.find('#search-box-clear').simulate('click');

        expect(onClear).toHaveBeenCalled();
      });
      it('should trigger onFocus on input focus', () => {
        const onFocus = jest.fn();

        const wrapper = shallow(<SearchBox onFocus={onFocus} />);
        wrapper.find('input').simulate('focus');

        expect(onFocus).toHaveBeenCalled();
      });
      it('should trigger onInput on input change', () => {
        const onInput = jest.fn();
        const enteredInputValue = 'a';
        const wrapper = shallow(<SearchBox onInput={onInput} />);
        wrapper
          .find('input')
          .simulate('input', { target: { value: enteredInputValue } });

        expect(onInput).toHaveBeenCalled();
      });
      it('should trigger onKeyPress on input change', () => {
        const onKeyPress = jest.fn();

        const wrapper = shallow(<SearchBox onKeyPress={onKeyPress} />);
        wrapper.find('input').simulate('keyPress');

        expect(onKeyPress).toHaveBeenCalled();
      });
      it('should trigger onSubmit on input change', () => {
        const onSubmit = jest.fn();

        const wrapper = shallow(<SearchBox onSubmit={onSubmit} />);
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });

        expect(onSubmit).toHaveBeenCalled();
      });
    });
  });
});
