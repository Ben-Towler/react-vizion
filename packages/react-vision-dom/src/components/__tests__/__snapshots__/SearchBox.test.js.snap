// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox applies its default props 1`] = `
<div
  className="cvi-SearchBox"
>
  <form
    onSubmit={[Function]}
  >
    <div>
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        className="cvi-SearchBox-input"
        id="search-box-input"
        onInput={[Function]}
        spellCheck={false}
        type="search"
      />
      <div>
        <span
          className=""
        />
      </div>
      <span
        className="cvi-SearchBox-clear"
        id="search-box-clear"
        onClick={[Function]}
      />
      <button
        className="cvi-SearchBox-submit"
        type="submit"
      />
    </div>
  </form>
</div>
`;
